spec.html

```diff
@@ -32544,67 +32544,54 @@ <h1>Array.prototype.some ( _callbackfn_ [ , _thisArg_ ] )</h1>
       <emu-clause id="sec-array.prototype.sort">
         <h1>Array.prototype.sort ( _comparefn_ )</h1>
         <p>The elements of this array are sorted. The sort must be stable (that is, elements that compare equal must remain in their original order). If _comparefn_ is not *undefined*, it should be a function that accepts two arguments _x_ and _y_ and returns a negative value if _x_ &lt; _y_, zero if _x_ = _y_, or a positive value if _x_ &gt; _y_.</p>
-        <p>Upon entry, the following steps are performed to initialize evaluation of the `sort` function:</p>
+        <p>The following steps are taken:</p>
         <emu-alg>
           1. If _comparefn_ is not *undefined* and IsCallable(_comparefn_) is *false*, throw a *TypeError* exception.
           1. Let _obj_ be ? ToObject(*this* value).
           1. Let _len_ be ? ToLength(? Get(_obj_, `"length"`)).
-        </emu-alg>
-        <p>Within this specification of the `sort` method, an object, _obj_, is said to be <em>sparse</em> if the following algorithm returns *true*:</p>
-        <emu-alg>
-          1. For each integer _i_ in the range 0 &le; _i_ &lt; _len_, do
-            1. Let _elem_ be _obj_.[[GetOwnProperty]](! ToString(_i_)).
-            1. If _elem_ is *undefined*, return *true*.
-          1. Return *false*.
+          1. Let _items_ be a new empty List.
+          1. Let _k_ be 0.
+          1. Let _undefinedCount_ be 0.
+          1. Repeat, while _k_ &lt; _len_
+            1. Let _Pk_ be ! ToString(_k_).
+            1. Let _kPresent_ be ? HasProperty(_obj_, _Pk_).
+            1. If _kPresent_ is *true*, then
+              1. Let _kValue_ be ? Get(_obj_, _Pk_).
+              1. If _kValue_ is *undefined*, then
+                1. Set _undefinedCount_ to _undefinedCount_ + 1.
+              1. Else,
+                1. Append _kValue_ to _items_.
+            1. Set _k_ to _k_ + 1.
+          1. Let _itemCount_ be the length of _items_.
+          1. Sort _items_ using an implementation-dependent sequence of calls to SortCompare.
+          1. Let _j_ be 0.
+          1. Repeat, while _j_ &lt; _itemCount_
+            1. Perform ? Set(_obj_, ! ToString(_j_), items[_j_], *true*).
+            1. Set _j_ to _j_ + 1.
+          1. Repeat, while _j_ &lt; _itemCount_ + _undefinedCount_
+            1. Perform ? Set(_obj_, ! ToString(_j_), *undefined*, *true*).
+            1. Set _j_ to _j_ + 1.
+          1. Repeat, while _j_ &lt; _len_
+            1. Let _Pj_ be ! ToString(_j_).
+            1. Let _jPresent_ be ? HasOwnProperty(_obj_, _Pj_).
```

Comment:
It should be possible to omit this call to `HasOwnProperty` and unconditionally call `DeletePropertyOrThrow`. (This matches what SpiderMonkey does [here](https://searchfox.org/mozilla-central/rev/da855d65d1fbdd714190cab2c46130f7422f3699/js/src/builtin/Sorting.js#263-268) and [here](https://searchfox.org/mozilla-central/rev/da855d65d1fbdd714190cab2c46130f7422f3699/js/src/builtin/Sorting.js#263-268).)

Reply:
I checked the spec of `DeletePropertyOrThrow` and the `delete` syntax (used by SpiderMoneky), and we should indeed be able to remove the `HasOwnProperty` check. Any other opinions?

@ljharb

Reply:
What if `obj` is a proxy - wouldn't removing this "has" check be observable?

It does seem, though, like `obj.[[Delete]](P)` will return `true` if the property was never present to begin with, which suggests the check can be removed.

Reply:
I removed the `HasOwnProperty` check. The corresponding [V8 CL](https://crrev.com/c/1809366) landed yesterday.